File: app/src/main/java/com/app/missednotificationsreminder/binding/model/ApplicationsSelectionViewModel.java
package com.app.missednotificationsreminder.binding.model;

import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;

import com.app.missednotificationsreminder.data.model.ApplicationItem;
import com.app.missednotificationsreminder.data.model.NotificationData;
import com.app.missednotificationsreminder.data.model.util.ApplicationIconHandler;
import com.app.missednotificationsreminder.di.qualifiers.IoThreadScheduler;
import com.app.missednotificationsreminder.di.qualifiers.MainThreadScheduler;
import com.app.missednotificationsreminder.di.qualifiers.SelectedApplications;
import com.app.missednotificationsreminder.ui.view.ApplicationsSelectionView;
import com.app.missednotificationsreminder.ui.widget.ApplicationsSelectionAdapter;
import com.f2prateek.rx.preferences.Preference;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import javax.inject.Inject;

import rx.Observable;
import rx.Scheduler;
import rx.Single;
import rx.functions.Action1;
import rx.functions.Func1;
import rx.subjects.PublishSubject;
import timber.log.Timber;

/**
 * The view model for the applications selection view
 *
 * @author Eugene Popovich
 */
public class ApplicationsSelectionViewModel extends BaseViewModel {
    private ApplicationsSelectionView mView;
    private Preference<Set<String>> mSelectedApplications;
    private Scheduler mMainThreadScheduler;
    private Scheduler mIoThreadScheduler;
    private PackageManager mPackageManager;
    Observable<List<NotificationData>> mNotificationDataObservable;

    private PublishSubject<Set<String>> dataLoadSubject;

    /**
     * @param view                 the related view
     * @param selectedApplications preference to store/retrieve selected applications
     * @param mainThreadScheduler
     * @param ioThreadScheduler
     * @param packageManager
     */
    @Inject public ApplicationsSelectionViewModel(ApplicationsSelectionView view, @SelectedApplications Preference<Set<String>> selectedApplications,
                                                  Observable<List<NotificationData>> notificationDataObservable,
                                                  @MainThreadScheduler Scheduler mainThreadScheduler,
                                                  @IoThreadScheduler Scheduler ioThreadScheduler,
                                                  PackageManager packageManager) {
        this.mMainThreadScheduler = mainThreadScheduler;
        this.mIoThreadScheduler = ioThreadScheduler;
        this.mView = view;
        this.mPackageManager = packageManager;
        this.mSelectedApplications = selectedApplications;
        mNotificationDataObservable = notificationDataObservable;
        init();
    }

    private void init() {

        // initialize data loading
        dataLoadSubject = PublishSubject.create();

        Observable<List<ApplicationItem>> result = dataLoadSubject //
                .flatMapSingle(this::packagesList) // load data
                .observeOn(mMainThreadScheduler) //
                .share();
        monitor(result //
                .subscribe(mView.getListLoadedAction(), errorHandler));
    }

    private Single<List<ApplicationItem>> packagesList(Set<String> selectedPreferences) {
        return mNotificationDataObservable
                .take(1)
                .toSingle()
                .map(ApplicationsSelectionAdapter::getNotificationCountData)
                .flatMap(notificationsCountInfo -> Single.fromCallable(() -> {
                    List<ApplicationItem> result = new ArrayList<>();
                    List<PackageInfo> packages = mPackageManager.getInstalledPackages(0);
                    for (PackageInfo packageInfo : packages) {
                        boolean selected = selectedPreferences.contains(packageInfo.packageName);
                        result.add(new ApplicationItem.Builder()
                                .checked(selected)
                                .applicationName(packageInfo.applicationInfo.loadLabel(mPackageManager))
                                .packageName(packageInfo.packageName)
                                .activeNotifications(notificationsCountInfo.containsKey(packageInfo.packageName) ? notificationsCountInfo.get(packageInfo.packageName) : 0)
                                .iconUri(new Uri.Builder()
                                        .scheme(ApplicationIconHandler.SCHEME)
                                        .authority(packageInfo.packageName)
                                        .build())
                                .build());
                    }
                    return result;
                }).subscribeOn(mIoThreadScheduler));
    }

    /**
     * Load the application data to the view
     */
    public void loadData() {
        mView.setLoadingState();
        dataLoadSubject.onNext(mSelectedApplications.get());
    }


    /**
     * The action to handle error which may occur during data loading operation
     */
    private final Action1<Throwable> errorHandler = t -> {
        Timber.e(t, "Unexpected error");
        mView.setErrorState();
    };
}

File: app/src/main/java/com/app/missednotificationsreminder/binding/model/ApplicationsSelectionViewModel.kt
package com.app.missednotificationsreminder.binding.model

import android.content.pm.PackageManager
import android.net.Uri
import androidx.lifecycle.LiveData
import androidx.lifecycle.viewModelScope
import com.app.missednotificationsreminder.data.model.ApplicationItem
import com.app.missednotificationsreminder.data.model.NotificationData
import com.app.missednotificationsreminder.data.model.util.ApplicationIconHandler
import com.app.missednotificationsreminder.di.qualifiers.SelectedApplications
import com.app.missednotificationsreminder.ui.widget.ApplicationsSelectionAdapter
import com.app.missednotificationsreminder.util.asFlow
import com.app.missednotificationsreminder.util.livedata.NonNullMutableLiveData
import com.f2prateek.rx.preferences.Preference
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.*
import rx.Observable
import timber.log.Timber
import java.util.*
import javax.inject.Inject

/**
 * The view model for the applications selection view
 *
 * @author Eugene Popovich
 */
class ApplicationsSelectionViewModel @Inject constructor(
        @param:SelectedApplications private val selectedApplicationsPref: Preference<Set<String>>,
        private val notificationDataObservable: Observable<List<NotificationData>>,
        private val packageManager: PackageManager) : BaseViewModel() {
    private val _viewState = NonNullMutableLiveData(ViewState(LoadingStatus.NotStarted, Collections.emptyList()))
    val viewState: LiveData<ViewState> = _viewState

    init {
        // initialize data loading
    }

    /**
     * Load the application data to the view
     */
    @ExperimentalCoroutinesApi
    fun loadData() {
        Timber.d("loadData: thread=%s", Thread.currentThread().name)
        if (_viewState.value.loadingStatus == LoadingStatus.Loading) {
            Timber.d("loadData: already loading, return")
            return
        }
        _viewState.value = _viewState.value.copy(loadingStatus = LoadingStatus.Loading)
        notificationDataObservable
                .asFlow()
                .take(1)
                .map { ApplicationsSelectionAdapter.getNotificationCountData(it) }
                .map { notificationsCountInfo ->
                    val result: MutableList<ApplicationItem> = ArrayList()
                    val packages = packageManager.getInstalledPackages(0)
                    val selectedApplications = selectedApplicationsPref.get() ?: emptySet()
                    for (packageInfo in packages) {
                        val selected = selectedApplications.contains(packageInfo.packageName)
                        result.add(ApplicationItem.Builder()
                                .checked(selected)
                                .applicationName(packageInfo.applicationInfo.loadLabel(packageManager))
                                .packageName(packageInfo.packageName)
                                .activeNotifications(if (notificationsCountInfo.containsKey(packageInfo.packageName)) notificationsCountInfo[packageInfo.packageName]!! else 0)
                                .iconUri(Uri.Builder()
                                        .scheme(ApplicationIconHandler.SCHEME)
                                        .authority(packageInfo.packageName)
                                        .build())
                                .build())
                    }
                    result.toList()
                }
                .flowOn(Dispatchers.IO)
                .catch { t ->
                    Timber.e(t, "Unexpected error")
                    _viewState.value = _viewState.value.copy(loadingStatus = LoadingStatus.Error)
                }
                .onEach {
                    _viewState.value = _viewState.value.copy(loadingStatus = LoadingStatus.NotStarted, data = it)
                }
                .launchIn(viewModelScope)
    }
}

data class ViewState(val loadingStatus: LoadingStatus, val data: List<ApplicationItem>)

sealed class LoadingStatus {
    object NotStarted : LoadingStatus()
    object Loading : LoadingStatus()
    object Error : LoadingStatus()
}

